lib/factories/default-eventbus.js

```
@@ -0,0 +1,373 @@
+ *
+ * - key_field
+ *      If given, the event's Kafka 'key' will be extracted from the event at this
+ *      dotted object path
+ *      Default: undefined
+ *
+ *  - partition_field
+ *      If given, the event's Kafka 'partition' will be extracted from the event at this
+ *      dotted object path.
+ *      Default: undefined
+ *
+ * - id_field
+ *      This field will be used as the event's 'id' in log messages
+ *      Default: undefined
+ *
+ * - kafka.conf
+ *      node-rdkafka KafkaProducer configuration
+ *
+ * - kafka.topic_conf
+ *      node-rdkafka KafkaProducer topic configuration
+ */
+
+
+const defaultOptions = {
+    schema_uri_field:       '$schema',
+    schema_base_uri:        undefined,
+    schema_file_extension:  undefined,
+    stream_field:           undefined,
+    topic_prefix:           undefined,
+    key_field:              undefined,
+    partition_field:        undefined,
+    id_field:               undefined,
+    kafka: {
+        conf: {
+            'metadata.broker.list': 'localhost:9092'
+        },
+        topic_conf: {}
+    },
+};
+
+/**
+ * Returns a function that extracts the event's schema URI.
+ * @param {Object} options
+ * @param {string} options.schema_uri_field
+ *      Used to extract the event's schema URI.
+ * @return {function(Object): string}
+ */
+function makeExtractSchemaUri(options) {
+    const schemaUriField = _.get(options, 'schema_uri_field', defaultOptions.schema_uri_field);
+    return (event) => {
+        try {
+            return objectGet(event, _.get(options, 'schema_uri_field'));
+        } catch (err) {
+            // Wrap PropertyNotFoundError in a new specific Error about missing Schema URI.
+            if (err instanceof PropertyNotFoundError) {
+                throw new EventSchemaUriMissingError(
+                    'Event schema URI cannot be extracted. ' +
+                    `Event must have a '${schemaUriField}' property`
+                );
+            }
```

And if any other error - just swallow?

---

Oops, no this is supposed to rethrow...used to using Promise.catch where I can catch only a specific type.